### PR TITLE
Feat/pupil 250/lesson engine provider

### PR DESCRIPTION
**Pull request by kimon-satan**
### Story
Implement the lesson engine provider and implement user flow in line with figma designs using placeholder views

### Scope
- the scope of this ticket is the user flow and lesson mechanics
- please consider all rendering out of scope for this ticket. The majority of components will be replaced in future tickets.

### Caveats
- there are no tests for views as they are currently placeholder . They have been excluded from sonar cloud for now

### Implementation
- please look at src/hooks/useStateCallback.ts and check that this hook is a satisfactory approach for taking actions after state has updated.

### Testing
Go to any non-legacy lesson eg.
https://deploy-preview-2188--oak-web-application.netlify.thenational.academy/pupils/programmes/combined-science-secondary-ks4-foundation-aqa/units/measuring-waves/lessons/representing-longitudinal-waves
Work through the entire useflow checking behaviour is as expected
You can also jump straight to a section by using a query flag eg. 
https://deploy-preview-2188--oak-web-application.netlify.thenational.academy/pupils/programmes/combined-science-secondary-ks4-foundation-aqa/units/measuring-waves/lessons/representing-longitudinal-waves?section=exit-quiz

### Acceptance criteria

- [x]  state for sections completed
- [x]  state for currentSection
- [x]  Adapt pupil page to render appropriate section based on state
- [x]  create stub views for: lessonOverview, lessonIntro, video, lessonReview
- [x]  flow returns to overview on section complete (except final section which directs to review)
- [x]  completed sections marked on overview page
- [x] proceed button navigates to first incomplete section
- [x] direct navigation to sections from overview page
- [x] quiz results visible on overview and review pages
- [x] back button returns to overview page

### CODE REVIEW

**Review by rightsaidjames — COMMENTED** [review]
Looks good - I can complete the lesson in any order I like (or just go with the flow by clicking the 'proceed to next section' button on the overview page), and if I complete the  lesson in the 'wrong order then the 'proceed to next section' button will take me to the next incomplete section of the lesson.

The only thing that stood out as possible business logic errors are:
* If you exit a quiz before you complete it, it will give you a score out of 6 based on your current progress (e.g. 0/6 if you answer 1 question wrong then exit the quiz). I suggest we park this for now and revisit it when we have more of the UI and business logic implemented.
* If you complete the lesson in the 'wrong' order, the lesson review page will display the lesson sections in the wrong order. We agreed in the standup that this was a bug that should be fixed pre-merge.

Screenshot 2024-01-18 at 09 02 12
Screenshot 2024-01-18 at 09 04 29
Screenshot 2024-01-18 at 09 04 41
Screenshot 2024-01-18 at 09 16 24

File: src/hooks/useStateCallback.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    // cb.current is `undefined` on initial render,
+    // so we only invoke callback on state *updates*
+    if (cbRef.current) {
+      cbRef.current(state);
+      cbRef.current = undefined; // reset callback after execution
+    }
+  }, [state]);
+
+  return [state, setStateCallback];
+}
```
Comment (carlmw) [inline src/hooks/useStateCallback.ts:35]: I feel like this is a workaround to a problem that could be solved by consolidating the state in the provider.

**Review by carlmw — COMMENTED** [review]
Some reckons on my part, but I feel like that new utility hook is unnecessary

**Comment by rightsaidjames** [comment]
Another edge case, could potentially be parked for now - if a lesson doesn't have an exit quiz, it still appears on the overview page and, because you can't complete a non-existent exit quiz, you can never complete the lesson:
https://deploy-preview-2188--oak-web-application.netlify.thenational.academy/pupils/programmes/english-primary-ks1/units/writing-lower-case-letters-in-print/lessons/formation-of-i-l-and-t
Screenshot 2024-01-18 at 11 08 22
Screenshot 2024-01-18 at 11 08 27

File: src/pages/pupils/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug]/index.tsx [hunk 2]
```diff
@@ -1,48 +1,78 @@
...
-  const { starterQuiz } = curriculumData;
+  const { currentSection, updateCurrentSection } = useLessonEngineContext();
+  const searchParams = useSearchParams();
+  const [overrideApplied, setOverrideApplied] = useState(false);
+
+  const overrideSection = searchParams.get("section");
+  if (overrideSection && isLessonSection(overrideSection) && !overrideApplied) {
+    setOverrideApplied(true);
+    updateCurrentSection(overrideSection);
+  }
```
[the remark this answers is not in the document]
Comment (kimon-satan) [reply src/pages/pupils/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug]/index.tsx:46]: This one is temporary code to allow people to jump to sections so I prefer to leave it at the page level. I'll add a comment to that effect.

File: src/hooks/useStateCallback.ts (hunk 1 above)
Comment (kimon-satan) [reply src/hooks/useStateCallback.ts:35]: I'm not sure I understand what you mean by consolidating the state here. Perhaps we need to do a call for you to explain this one ?

**Comment by kimon-satan** [comment]
> Another edge case, could potentially be parked for now - if a lesson doesn't have an exit quiz, it still appears on the overview page and, because you can't complete a non-existent exit quiz, you can never complete the lesson: https://deploy-preview-2188--oak-web-application.netlify.thenational.academy/pupils/programmes/english-primary-ks1/units/writing-lower-case-letters-in-print/lessons/formation-of-i-l-and-t Screenshot 2024-01-18 at 11 08 22 Screenshot 2024-01-18 at 11 08 27

Damn it ! Could you provide me with some links where there is no exit quiz. Lets make this another ticket in any case.

File: src/components/PupilComponents/LessonEngineProvider/LessonEngineProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  proceedToNextSection: () => void;
+  updateQuizResult: (vals: { grade: number; maxScore: number }) => void;
+} | null;
+
+export const LessonEngineContext = createContext<LessonEngineContextType>(null);
+
+export const useLessonEngineContext = () => {
+  const context = useContext(LessonEngineContext);
+  if (!context) {
+    throw new Error("`QuizEngineProvider` is not available");
```
[the remark this answers is not in the document]
Comment (kimon-satan) [reply src/components/PupilComponents/LessonEngineProvider/LessonEngineProvider.tsx:42]: I'm not really sure how this case is different from this one which you recommended in a previous review
`export const useQuizEngineContext = () => {
  const context = useContext(QuizEngineContext);
  if (!context) {
    throw new Error("`QuizEngineProvider` is not available");
  }
  return context;
};`

File: src/components/PupilComponents/LessonEngineProvider/LessonEngineProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        if (
+          _completedSections.length ===
+          lessonSections.filter((s) => s !== "overview" && s !== "review")
+            .length
+        ) {
+          setCurrentSection("review");
+        } else setCurrentSection("overview");
+      },
+    );
+  };
```
[the remark this answers is not in the document]
Comment (kimon-satan) [reply src/components/PupilComponents/LessonEngineProvider/LessonEngineProvider.tsx:79]: happy to refactor / comment but currentSection is only occasionally dependent on completed section as you can also freely navigate.

File: src/components/PupilComponents/QuizEngineProvider/QuizEngineProvider.tsx [hunk 5]
```diff
@@ -170,14 +181,26 @@ export const QuizEngineProvider = memo((props: QuizEngineProps) => {
...
+    ],
   );
 
   const handleNextQuestion = useCallback(() => {
-    setCurrentQuestionIndex((prev) => Math.min(prev + 1, maxScore));
-  }, [maxScore]);
+    setCurrentQuestionIndex(
+      (prev) => Math.min(prev + 1, maxScore),
+      (index) => {
+        if (index === maxScore) {
```
[the remark this answers is not in the document]
Comment (kimon-satan) [reply src/components/PupilComponents/QuizEngineProvider/QuizEngineProvider.tsx:198]: fair enough - will change to numQuestions

**Review by carlmw — APPROVED** [review]
🚀

**Pull request merged**